### PR TITLE
kvstore: switch to kvstore v2 hostcalls

### PR DESCRIPTION
**Pull request by dgryski**
This doesn't expose any new API surface but does switch the kvstore backend hostcalls to the v2 API.

### CODE REVIEW

File: internal/abi/fastly/kvstore_guest.go [hunk 1]
```diff
@@ -10,166 +10,343 @@ import (
...
+	config prim.Pointer[kvLookupConfig],
+	lookupHandle prim.Pointer[kvstoreLookupHandle],
+) FastlyStatus
+
+// Lookup returns a handle to a pending lookup operation
+func (kv *KVStore) Lookup(key string) (kvstoreLookupHandle, error) {
+	keyBuffer := prim.NewReadBufferFromString(key).Wstring()
+
+	// empty
+	var opts kvLookupConfigMask
```
Comment (cee-dub) [inline internal/abi/fastly/kvstore_guest.go:101]:
```suggestion
	var mask kvLookupConfigMask
```
For consistency?

File: internal/abi/fastly/kvstore_guest.go [hunk 2]
```diff
@@ -10,166 +10,343 @@ import (
...
 ) FastlyStatus
 
-// Lookup returns the value for key, if it exists.
-func (o *KVStore) Lookup(key string) (io.Reader, error) {
+// LookupWait returns a lookup response for a pending lookup handle
+func (kv *KVStore) LookupWait(lookuph kvstoreLookupHandle) (KVLookupResult, error) {
 	body := HTTPBody{h: invalidBodyHandle}
 
-	keyBuffer := prim.NewReadBufferFromString(key).Wstring()
+	meta := prim.NewWriteBuffer(kvstoreMetadataMaxBufLen)
```
Comment (cee-dub) [inline internal/abi/fastly/kvstore_guest.go:151]: Can we reduce garbage here by reusing a copy buffer of this length, similar to changes we made in HTTP for project Isetta?

File: internal/abi/fastly/kvstore_guest.go [hunk 3]
```diff
@@ -10,166 +10,343 @@ import (
...
-	if err := fastlyObjectStoreLookup(
-		o.h,
-		keyBuffer.Data, keyBuffer.Len,
+	var kvErr KVError = KVErrorUninitialized
+
+	if err := fastlyKVStoreLookupWait(
+		lookuph,
 		prim.ToPointer(&body.h),
+		prim.ToPointer(meta.U8Pointer()),
+		meta.Cap(),
```
Comment (cee-dub) [inline internal/abi/fastly/kvstore_guest.go:160]:
```suggestion
		prim.ToPointer(meta.U8Pointer()), meta.Cap(),
```
To keep line groupings as in the `func` declaration.

File: internal/abi/fastly/kvstore_guest.go [hunk 4]
```diff
@@ -10,166 +10,343 @@ import (
...
+
+//go:wasmimport fastly_kv_store insert_wait
+//go:noescape
+func fastlyKVStoreInsertWait(
+	h kvstoreInsertHandle,
+	kvErr prim.Pointer[KVError],
+) FastlyStatus
+
+// InsertWait returns the status of the given pending insertion handle.
+func (kv *KVStore) InsertWait(inserth kvstoreInsertHandle) error {
```
Comment (cee-dub) [inline internal/abi/fastly/kvstore_guest.go:256]: Total 🚲 🏠:

```suggestion
func (kv *KVStore) InsertWait(insertH kvstoreInsertHandle) error {
```
Capitalizing the H looks less weird to me.

File: internal/abi/fastly/types.go [hunk 5]
```diff
@@ -500,8 +501,177 @@ const (
...
+type (
+	kvstoreHandle       handle
+	kvstoreLookupHandle handle
+	kvstoreInsertHandle handle
+	kvstoreDeleteHandle handle
+	kvstoreListHandle   handle
+)
+
+const (
+	invalidKVStoreHandle  = kvstoreHandle(math.MaxUint32 - 1)
```
Comment (cee-dub) [inline internal/abi/fastly/types.go:518]: Sorry for not knowing why these are 1 less than the max as the sentinel value. Why is that?

**Review by cee-dub — APPROVED** [review]
Thanks for the updates to this! Nothing blocking in my review, so adjust as you see fit.

File: internal/abi/fastly/kvstore_guest.go (hunk 4 above)
Comment (dgryski) [reply internal/abi/fastly/kvstore_guest.go:256]: Then I should probably go and make the the rest of them `H`-andle too.

File: internal/abi/fastly/types.go (hunk 5 above)
Comment (dgryski) [reply internal/abi/fastly/types.go:518]: To match the Rust SDK
```
fastly-shared/src/lib.rs
23:pub const INVALID_KV_PENDING_LOOKUP_HANDLE: u32 = std::u32::MAX - 1;
24:pub const INVALID_KV_PENDING_INSERT_HANDLE: u32 = std::u32::MAX - 1;
25:pub const INVALID_KV_PENDING_DELETE_HANDLE: u32 = std::u32::MAX - 1;
26:pub const INVALID_KV_PENDING_LIST_HANDLE: u32 = std::u32::MAX - 1;
29:pub const INVALID_KV_STORE_HANDLE: u32 = std::u32::MAX - 1;
```

File: internal/abi/fastly/kvstore_guest.go (hunk 1 above)
Comment (dgryski) [reply internal/abi/fastly/kvstore_guest.go:101]: Yup.

**Comment by dgryski** [comment]
Merging.  Will put up a second PR to expand the API surface to include the new v2 features.

**Pull request merged**